test/common/buffer/owned_impl_test.cc

```diff
@@ -1477,6 +1461,67 @@ TEST(BufferMemoryAccountTest, LinearizedBufferShouldChargeItsAssociatedAccount)
   EXPECT_EQ(buffer_three_account->balance(), 8192);
 }
 
+TEST(BufferMemoryAccountTest, ManagesAccountBalanceWhenPrepending) {
+  auto prepend_to_account = std::make_shared<BufferMemoryAccountImpl>();
+  Buffer::OwnedImpl buffer_to_prepend_to(prepend_to_account);
+  ASSERT_EQ(prepend_to_account->balance(), 0);
+
+  auto prepend_account = std::make_shared<BufferMemoryAccountImpl>();
+  Buffer::OwnedImpl buffer_to_prepend(prepend_account);
+  ASSERT_EQ(prepend_account->balance(), 0);
+
+  Buffer::OwnedImpl unowned_buffer_to_prepend;
+
+  unowned_buffer_to_prepend.add("World");
+  buffer_to_prepend.add("Goodbye World");
+  EXPECT_EQ(prepend_account->balance(), 4096);
+
+  // Prepend the buffers.
+  buffer_to_prepend_to.prepend(buffer_to_prepend);
+  EXPECT_EQ(prepend_account->balance(), 4096);
+  EXPECT_EQ(prepend_to_account->balance(), 0);
+
+  buffer_to_prepend_to.prepend(unowned_buffer_to_prepend);
+  EXPECT_EQ(prepend_to_account->balance(), 4096);
+
+  // Prepend a string view.
+  buffer_to_prepend_to.prepend("Hello ");
+  EXPECT_EQ(prepend_to_account->balance(), 8192);
+}
+
+TEST(BufferMemoryAccountTest, ExtractingSliceWithExistingStorageCreditsAccountOnce) {
+  auto buffer_account = std::make_shared<BufferMemoryAccountImpl>();
+  Buffer::OwnedImpl buffer(buffer_account);
+  ASSERT_EQ(buffer_account->balance(), 0);
+
+  buffer.appendSliceForTest("Slice 1");
+  buffer.appendSliceForTest("Slice 2");
+  EXPECT_EQ(buffer_account->balance(), 8192);
+
+  // Account should only be credited when slice is extracted.
+  // Not on slice dtor.
+  {
+    auto slice = buffer.extractMutableFrontSlice();
+    EXPECT_EQ(buffer_account->balance(), 4096);
+  }
+
+  EXPECT_EQ(buffer_account->balance(), 4096);
+}
+
+TEST(BufferMemoryAccountTest, UnusedReservationSlicesShouldBeCreditedAccountAfterCommit) {
+  auto buffer_account = std::make_shared<BufferMemoryAccountImpl>();
+  Buffer::OwnedImpl buffer(buffer_account);
+  ASSERT_EQ(buffer_account->balance(), 0);
+
+  // Create an excessively large reservation. The unused portion should be
+  // credited on destruction.
+  auto reservation = buffer.reserveForRead();
+  EXPECT_EQ(buffer_account->balance(), 16384 * 8);
```

Comment:
Charging the account on reserve seems a bit odd, but I guess it's ok.  Waiting until commit would allow us to avoid increments to the account shared_ptr until we know that the reserved slices are being committed into the buffer.

Reply:
The memory is already allocated on reserve. The commit just says I'm using X bytes which translates to some Y slices of it.

We could have the slices reserved untagged but that would add additional complexity of needing to remove accounts from empty slices that get used in the reservation and then some memory that will get added to buffers will go unaccounted for some time.

Reply:
The buffer doesn't really own the slices in the reservation until commit happens.  The time between reserve and commit is always short.  Let's do as you suggest: not set account on reserve and set it on commit instead.

Reply:
Done.